src/protosym/core/sym.py

```diff
@@ -0,0 +1,615 @@
+
+    >>> import math
+    >>> from protosym.core.sym import Sym, PyOp1, PyFunc1
+    >>> Integer = Sym.new_atom('Integer', int)
+    >>> Function = Sym.new_atom('Function', int)
+    >>> cos = Function('cos').rep
+    >>> a = Sym.new_wild('a')
+    >>> ar = a.rep
+    >>> one = Integer(1).rep
+
+    Make some symbolically typed functions for evaluation:
+
+    >>> f64_from_int = PyFunc1(float)
+    >>> f64_cos = PyOp1(math.cos)
+
+    Now we can make an :class:`Evaluator` and add rules to it:
+
+    >>> eval_f64 = Sym.new_evaluator('eval_f64', float)
+    >>> eval_f64
+    eval_f64
+    >>> eval_f64[Integer[a]] = f64_from_int(a)
+    >>> eval_f64[Sym(cos(ar))] = f64_cos(a)
+
+    Now make an expression and evaluate the expression:
+
+    >>> cos_one = Sym(cos(one))
+    >>> print(cos_one)
+    cos(1)
+    >>> eval_f64(cos_one) # doctest: +ELLIPSIS
+    0.5403023058681...
+
+    See Also
+    ========
+
+    protosym.core.evaluate::Evaluator
+    protosym.simplecas::Expr.new_evaluator
+    """
+
+    def __init__(self, name: str):
+        """Create a new SymEvaluator."""
+        self.name = name
+        self.evaluator = Evaluator[T_val]()
+
+    # e.g. eval_f64[cos(a)] = f64_cos(a)
+    @overload
+    def __setitem__(  # noqa: D105
+        self,
+        pattern: T_sym,
+        call: WildCall[T_sym, PyOp1[T_val]]
+        | WildCall[T_sym, PyOp2[T_val]]
+        | WildCall[T_sym, PyOpN[T_val]],
+    ) -> None:
+        ...
+
+    # e.g. eval_f64[Integer[a]] = f64_from_int
+    @overload
+    def __setitem__(  # noqa: D105
+        self,
+        pattern: SymAtomValue[T_sym, S_val],
+        call: WildCall[T_sym, PyFunc1[S_val, T_val]],
+    ) -> None:
+        ...
+
+    # e.g. eval_repr[AtomRule[a]] = AtomFunc(repr)
+    @overload
+    def __setitem__(  # noqa: D105
+        self,
+        pattern: AtomRuleType[T_sym],
+        call: WildCall[T_sym, AtomFunc[T_val]],
+    ) -> None:
+        ...
+
+    # e.g. eval_repr[HeadRule(a, b)] = HeadOp(...)
+    @overload
+    def __setitem__(  # noqa: D105
+        self,
+        pattern: HeadRuleType[T_sym],
+        call: WildCall[T_sym, HeadOp[T_val]],
+    ) -> None:
+        ...
+
+    def __setitem__(  # noqa: C901
+        self,
+        pattern: T_sym
+        | SymAtomValue[T_sym, S_val]
+        | AtomRuleType[T_sym]
+        | HeadRuleType[T_sym],
+        call: WildCall[T_sym, PyOp1[T_val]]
+        | WildCall[T_sym, PyOp2[T_val]]
+        | WildCall[T_sym, PyOpN[T_val]]
+        | WildCall[T_sym, PyFunc1[S_val, T_val]]
+        | WildCall[T_sym, AtomFunc[T_val]]
+        | WildCall[T_sym, HeadOp[T_val]],
+    ) -> None:
+        """Add an evaluation rule."""
+        if not isinstance(call, WildCall):
+            raise BadRuleError("Rule function should be a symbolic call.")
+
+        if isinstance(call.op, PyOpN):
+            (callarg,) = call.args
+            if pattern.args != (star(callarg),):
+                raise BadRuleError("Nary function needs a star-rule.")
```

Comment:
Took me a second to understand what "Nary" was referring to here. Thoughts on using multary or multiary instead?

Reply:
Maybe `n-ary` is better?

I've never heard of multary or multiary.

Another possibility would be "var-args" which is more like "polyadic". Perhaps these are actually more accurate because it is not a fixed "n" here.

Reply:
I changed this to "Varargs" which is perhaps more well known to Python programmers at least...

Reply:
The [Wikipedia article on Arity](https://en.wikipedia.org/wiki/Arity) states that the term for "more than 2-ary" can be either multary or multiary. I've come across both multary and multiary alongside unary and binary in textbooks before.

Reply:
The article uses n-ary more than either of those terms but in any case what we have is actually what the article refers to as "variadic" i.e. the number of arguments is not any fixed n but can be different in each call.

I suppose it helps here to see the error message in context:
```python
In [7]: from protosym.simplecas import *

In [8]: eval_f64[Add(a)] = PyOpN(sum)(a)
...
BadRuleError: varargs function needs a star-rule.

In [9]: eval_f64[Add(star(a))] = PyOpN(sum)(a)  # ok
```
I'm sure the message can be improved but it's a bit tricky to get it exactly right because it runtime it is hard to figure out exactly what the user is doing or trying to do. The Evaluator does not intrinsically have any idea that `Add` itself *should* be variadic so it's difficult to say "Add is variadic and needs a star rule". The problem is that the rhs uses a PyOpN with parameter `a` and so the lhs needs to have `star(a)` somewhere.

Testing this out I see other cases that are accepted at runtime for example:
```python
In [10]: eval_f64[Add(a)] = PyOp1(sum)(a)  # should be an error
```
The type-checker would reject that because `PyOp1(sum)` implies a function that is not `T -> T`.
```python
from __future__ import annotations

from protosym.core.sym import PyOp1
from protosym.simplecas import eval_f64, Add, a

eval_f64[Add(a)] = PyOp1(sum)(a)
```
Here mypy gives:
```
t.py:6:20: error: Cannot infer type argument 1 of "PyOp1"  [misc]
    eval_f64[Add(a)] = PyOp1(sum)(a)
                       ^~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```
It can't infer the type because `PyOp1` expects `Callable[[S], S]` and sum is like `Callable[[Iterable[T]], T]` and it is not clear how to find an `S` that unifies `Iterable[T]` and `T`.

I saw a few cases like this where I thought that the type-checker was just not smart enough and tried using `type: ignore`. Sure enough when I ran the tests and traced the failures back it came to right where mypy had been complaining.